src/fairseq2/recipes/lm/preference_finetune.py

```diff
@@ -0,0 +1,439 @@
+
+        if dp_gang.rank == 0:
+            init_device = root_gang.device
+
+        model = load_model(
+            model_card, gangs=gangs, device=init_device, dtype=torch.float32
+        )
+
+        root_gang.barrier()
+
+        log.info("Model loaded on data parallel rank 0.")
+
+    if not isinstance(model, DecoderModel):
+        raise ValueError("`config.model` must specify a decoder model.")
+
+    checkpoint_manager.save_model_metadata(
+        base_asset=model_card.name, family=model.family
+    )
+
+    dp_model = to_data_parallel(
+        model,
+        dp_gang,
+        config.data_parallelism,
+        log,
+        fsdp_skip_init=True,
+        fsdp_broadcast_state=not has_checkpoint,
+        fsdp_reshard_after_forward=config.fsdp_reshard_after_forward,
+        fsdp_mixed_precision_dtype=config.dtype,
+        fsdp_fp32_reduce=True,
+        fsdp_wrap_granularity=config.fsdp_wrap_granularity,
+    )
+
+    if config.activation_checkpointing:
+        use_layerwise_activation_checkpointing(dp_model)
+
+    if config.torch_compile:
+        model.decoder = compile_model(model.decoder, log)
+
+    # TODO(balioglu): investigate!
+    # The memory efficient SDPA implementation in PyTorch is not stable when
+    # used with padded inputs.
+    enable_memory_efficient_torch_sdpa(dp_model, False)
+
+    log_model(dp_model, log, rank=root_gang.rank)
+
+    # Load the reference model.
+    def _get_reference_model(
+        criterion_config: PreferenceCriterionConfig,
+    ) -> Union[Module, None]:
+        try:
+            criterion_config.reference_model
+        except ValueError:
+            return None
+
+        reference_model_card = retrieve_asset_card(criterion_config.reference_model)
+
+        log.info("Loading {} reference model on data parallel rank 0 (per shard).", reference_model_card.name)  # fmt: skip
+
+        # TODO: figure out how to load the reference model onto its own gangs
+        reference_model = load_model(
+            reference_model_card,
+            gangs=gangs,
+            device=init_device,
+            dtype=criterion_config.reference_dtype,
+        )
+
+        root_gang.barrier()
+
+        log.info("Reference model loaded on data parallel rank 0.")
+
+        reference_model.eval()
+
+        freeze_parameters(reference_model)
+
+        dp_reference_model = to_data_parallel(
+            reference_model,
+            dp_gang,
+            config.data_parallelism,
+            log,
+            fsdp_skip_init=True,
+            fsdp_broadcast_state=not has_checkpoint,
+            fsdp_reshard_after_forward=config.fsdp_reshard_after_forward,
+            fsdp_mixed_precision_dtype=config.dtype,
+            fsdp_fp32_reduce=True,
+            fsdp_wrap_granularity=config.fsdp_wrap_granularity,
+        )
+
+        return dp_reference_model
+
+    dp_reference_model = _get_reference_model(config.criterion_config)
```

Comment:
This should be called within `create_preference_unit` as not all preference optimization techniques use a reference model:

```python
if config.criterion == "dpo":
  reference_model = load_reference_model(config.dpo.reference_model, ...)
  return DpoFinetuneUnit(dp_model, reference_model, ...)
```